ta/pkcs11/src/attributes.c

```diff
@@ -207,6 +207,48 @@ bool get_bool(struct obj_attrs *head, uint32_t attribute)
 	return bbool;
 }
 
+bool attributes_match_reference(struct obj_attrs *candidate,
+				struct obj_attrs *ref)
+{
+	size_t count = ref->attrs_count;
+	unsigned char *ref_attr = ref->attrs;
+	uint32_t rc = PKCS11_CKR_GENERAL_ERROR;
+
+	if (!ref->attrs_count) {
+		DMSG("Empty reference: no match");
+		return false;
+	}
+
+	for (count = 0; count < ref->attrs_count; count++) {
+		struct pkcs11_attribute_head pkcs11_ref = { };
+		void *found = NULL;
+		uint32_t size = 0;
+
+		TEE_MemMove(&pkcs11_ref, ref_attr, sizeof(pkcs11_ref));
+
+		if (pkcs11_is_attr_bool(pkcs11_ref.id)) {
+			bool bb_ref = get_bool(ref, pkcs11_ref.id);
+			bool bb_candidate = get_bool(candidate, pkcs11_ref.id);
```

Comment:
When looking at some session handling things I noticed that when using get_bool() there is possible problem in here.

If candidate does not have ref's attribute at all it gets value "false" and if ref's lookup value was "false" it matches.

Now is this a real problem or not I do not yet know. In theory we have proper default values if value is missing when creating an object.

But if one searches for specific entries and that particular attribute is from different object class then this might be problem.

It might be safest to do something like:
```
rc = get_attribute_ptr(candidate, pkcs11_ref.id,
			&found, &size);
if (rc || size != pkcs11_ref.size)
  return false;

if (pkcs11_is_attr_bool(pkcs11_ref.id))
...
else
...
```

I would porbably also rename "found" to be more like "(canditate_)value" or so.

Reply:
Good catch @vesajaaskelainen . To clarify if I understand it correctly,  I am giving an example below 
eg 
CK_ATTRIBUTE search_template[] = {
                { CKA_TOKEN, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
                { CKA_ENCRYPT,  &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) }, \
        };

Trying to search with above template on a data object should return no objects ideally as the template has the property CKA_ENCRYPT in it which is not a valid attribute for data object. Adding this check would return no objects for the above template.

If above understanding is correct, I would agree it is safe to add the check you mentioned above.

Reply:
As you said @vesajaaskelainen, the purpose here was to allow qiuck access to identified boolean attributes called boolprops, which covered the generic object standard boolean CKA_xxx. If we assume `get_attribute_ptr()` returns a valid byte pointer to the boolean value, there is no purpose calling `get_bool()`.

All in one, I would agree to simplify to:
```c
for (count = 0; count < ref->attrs_count; count++) {
	struct pkcs11_attribute_head pkcs11_ref = { };
	void *found = NULL;
	uint32_t size = 0;

	TEE_MemMove(&pkcs11_ref, ref_attr, sizeof(pkcs11_ref));

	rc = get_attribute_ptr(candidate, pkcs11_ref.id, &found, &size);
	if (rc || !found || size != pkcs11_ref.size ||
	    TEE_MemCompare(ref_attr + sizeof(pkcs11_ref), found, size))
		return false;

	ref_attr += sizeof(pkcs11_ref) + pkcs11_ref.size;
}

return true;
```